sources/packages/backend/libs/integrations/src/esdc-integration/application-changes-report/application-changes-report.processing.service.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  ApplicationChangesReportProcessingResult,
+  APPLICATION_CHANGES_FILE_NAME_TIMESTAMP_FORMAT,
+} from "./models/application-changes-report-integration.model";
+import { SystemUsersService } from "@sims/services";
+
+@Injectable()
+export class ApplicationChangesReportProcessingService {
+  private readonly esdcConfig: ESDCIntegrationConfig;
+  constructor(
+    private readonly applicationService: ApplicationService,
+    private readonly applicationChangesReportIntegrationService: ApplicationChangesReportIntegrationService,
+    private readonly systemUsersService: SystemUsersService,
+    private readonly studentAssessmentService: StudentAssessmentService,
+    config: ConfigService,
+  ) {
+    this.esdcConfig = config.esdcIntegration;
+  }
+
+  /**
+   * Generate application changes report for the applications which has at least one e-Cert sent
+   * and the application study dates have changed after the first e-Cert
+   * or after the last time the application was reported for study dates change
+   * through application changes report. Once generated upload the report to the ESDC directory
+   * in SFTP server.
+   * @param processSummary process summary.
+   * @returns processing result.
+   */
+  async processApplicationChanges(
+    processSummary: ProcessSummary,
+  ): Promise<ApplicationChangesReportProcessingResult> {
+    processSummary.info(
+      "Retrieving all application changes which were not reported already.",
+    );
+    const applicationChanges =
+      await this.applicationService.getDateChangeNotReportedApplications();
+    const applicationChangesCount = applicationChanges.length;
+    processSummary.info(
+      `Found ${applicationChangesCount} application changes.`,
+    );
+    // Build application changes report CSV file content.
+    const fileContent =
+      this.applicationChangesReportIntegrationService.createApplicationChangesReportFileContent(
+        applicationChanges,
+      );
+    // Create file name and build file path.
+    const { fileName, remoteFilePath } = this.createRequestFileName();
+    // Upload the file to SFTP location.
+    try {
+      await this.applicationChangesReportIntegrationService.uploadRawContent(
+        fileContent,
+        remoteFilePath,
+      );
+    } catch (error: unknown) {
+      // Translate to friendly error message.
+      const errorDescription = `Unexpected error uploading the file: ${remoteFilePath} to SFTP.`;
+      processSummary.error(errorDescription, error);
+      throw new Error(errorDescription, { cause: error });
+    }
+    processSummary.info(
+      `Application changes report with file name: ${fileName} has been uploaded successfully.`,
+    );
+    if (applicationChangesCount) {
+      const reportedAssessmentIds = applicationChanges.map(
+        (applicationChange) => applicationChange.currentAssessment.id,
+      );
+      await this.studentAssessmentService.updateReportedDate(
```

Comment:
Raising the question also for history, here we assume that if the file was uploaded but the DB update failed would be no problem to generate the file again. Is it correct?

Reply:
Yes. In that case the records failed to be updated in DB, it will be expected to be reported in the retry and potentially the application changes will be sent twice(of course with different file names).